Integrate Weaviate Engine, Local LTR, Local Engine

kosmikdc opened the pull request:
- Adds Weaviate as a fully supported engine
- Adds localized search engine for supporting text tagging and semantic knowledge graphs
- Improves ch6.18 edit distance length forumla
- Adds Google colab to ch14
- Removes extra search request id return field in Ch15 
- Standardize the cities data to conform to the Entities schema

kosmikdc committed:
Adds all Weaviate Collection, Engine, all spark and indexing logic for weaviate, Local LTR, embedded Solr for text tagging/skg and many other features/cleanings.
Verified core functionality works, and ch4 for weaviate.
Need to verify embedded solr and local LTR.

kosmikdc committed:
SolrCollection.write() to honor overwrite flag
Corrects weaviate schema fixing issues with many notebooks, LTR/signals still pending
Adds field_id to ltr interface
Correct 6.18 edit distance for min_length >= (edit distance can be up to 3, as opposed to just 3)
Corrects 15.18 return_fields to not have "id" twice

kosmikdc committed:
Merge branch 'main' into weaviate_updated

kosmikdc committed:
Merge branch 'main' into weaviate_updated

kosmikdc committed:
Merge branch 'main' into weaviate_updated

kosmikdc committed:
Corrects zk_url
Adds necessary dependencies for 5.1
Corrects weaviate schema boolean typing for simplicity
Isolates "id_field" for ltr.search_with_model preventing unnecessary abstraction changes

kosmikdc committed:
Weaviate: Persists id when loading collections into views
Adds user_product_cluster schema to weaviate
Applies fields to entities when creating the entity collection
Fixes weaviate get_logged_features() function signature
Corrects ch5-spacy requirements (from nested dependencies)
Removes dual indexing in ch5
Adds ch14 google colab
"local engine" instead of embedded engine
Correct cities loader to have surface/canonical

treygrainger commented on docker-compose.yml:
```diff
@@ -119,10 +146,19 @@ services:
       - all
       - ""
 
+  vespa:
```
@kosmikdc I'm assuming this isn't supposed to be in this branch.

treygrainger replied:
I also fixed the expose format to work with the Docker v29 release. Not sure if this is a regression in Docker (probably), but we don't use port ranges anyway, so just using the format in the standard for a single port to resolve.

https://github.com/treygrainger/ai-powered-search/commit/169f4af66671b539c6cb476c9c0ea1a19cfd04ce

treygrainger commented on engines/weaviate/build/spark-connector-1.4.0.jar:
@kosmikdc  - same as last comment. Was this jar file included intentionally?

kosmikdc replied:
After writing that yesterday I remembered that I custom built the 1.3.4a jar as the fix the geocoordinate indexing bug. I debugged the code and created a fix. That feature has sense been fixed in 1.4.0, but there's an error when indexing `java.lang.NoSuchMethodError: 'java.lang.Object scala.collection.immutable.Map$.apply(scala.collection.immutable.Seq)'
	at io.weaviate.spark.WeaviateOptions.<init>(WeaviateOptions.scala:49)`

I'll debug this a little as to figure out if I get get 1.4.0 working to use the remote jar, so I agree adding 30mb to the repo is undesirable and bad form.

treygrainger commented on docker-compose.yml:
@kosmikdc Due to whatever's going on with the docker compose, I can only run Solr (`docker compose up solr`) or Solr + Vespa (`docker compose up`) currently. Will try running Weaviate once it's working. Happy to hop on if you can't reproduce, but the current behavior's super weird.

I'm on the branch for this PR (`weaviate_updated`)

treygrainger replied:
Ok, I dug into this a bit. I think this is cause by Docker v29 (released two weeks ago) breaking support for port range parsing. This might be a bug in Docker compose that's been fixed but hasn't been released yet:
https://github.com/docker/compose/commit/6f365395e54a2670589c6850a6a4614c83f7ca8e

It works fine on Docker v28.

Someone in the [AI-Powered Search: Modern Retrieval for Humans & Agents](https://aipoweredsearch.com/live-course) course reported running into this issue on the `main` branch, so I'm going to prioritize getting it fixed, since I have no idea when and if Docker will revert the breaking change.

kosmikdc replied:
Glad you could find a root cause. I wasn't getting the errors as I hadn't updated docker yet.

treygrainger committed:
Merge in main

treygrainger committed:
Remove vespa from weaviate branch

treygrainger committed:
Fix expose port format to work with Docker v29 release

kosmikdc commented on engines/weaviate/build/spark-connector-1.3.4a.jar:
Removed

kosmikdc committed:
Improve logging
Improve local solr recovery
local_engine as name instead of embedded engine
Pull remote weaviate spark jar

kosmikdc committed:
Adds feature requirement to reviews dataset indexer
cleans up initialize engine signature
load weaviate spark plugin remotely

kosmikdc committed:
Merge branch 'main' into weaviate_updated

kosmikdc committed:
Removes all set_engine calls outside of the test suite.
Removes unnecessary spark binaries

kosmikdc committed:
Cleans up warnings in several chapters (no significant warnings were hidden)
Run solr across all notebooks, correcting notebooks with inconsistencies or failures (ch5, ch6, ch7, ch12, ch13, ch14, ch15)

kosmikdc committed:
Adds create_collection and load_index_time_boosting_dataframe to respective collection classes, simplifying index_time_boosts.py and spark/__init__.py
Local engine memory limit set to 512m
Solr create collection logging improvement
Neatens indexing logic for multi-engine indexing
Improves variable naming in aips/__init__.py
Removes unnecessary code in weaviatecollection, also correctly deletes the index when overwrite is used in write()

kosmikdc committed:
Merge branch 'main' into weaviate_updated

treygrainger commented on chapters/ch05/2.index-datasets.ipynb:
```diff
@@ -28,9 +28,53 @@
+   "execution_count": 2,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
Same with other collections in this chapter

kosmikdc committed:
Saved working directory and index state WIP on weaviate_updated: 5cb623c WeaviteCollection.write(): Correctly deletes, and recreates the collection when overwrite = True, accounting for the 'overwrite' bug in the weaviate spark plugin Local Engine: Fixes memory constraint that was pushed incorrectly, which was causing initialization to fail. SolrEngine.does_collection_exist improvements

kosmikdc committed:
Weaviate LTR to use correct ID field for products/movies

kosmikdc committed:
Adds semantic shutdown call in ch9 to prevent resource issues in the chapter

kosmikdc committed:
Refactors ch11.4 to be engine agnostic.
Removes fieldlength feature

kosmikdc committed:
Clean up healthcheck logs during semantic engine start

kosmikdc committed:
Adds notebook dependency to weaviate

kosmikdc committed:
Change dataset variable name in build_collection() to prevent overwriting of dataset var when dual indexing

kosmikdc committed:
Merge branch 'main' into weaviate_updated

kosmikdc committed:
Fixes title index null configuration for non-null filtering along with collection side logic to properly handle  (5.3)
Fixes case sensitivity issue preventing proper ID queries for WeaviateLTR (12.0, 12.1)

kosmikdc committed:
Weaviate: Removes popularity_vector, now just using stars_rating.
Weaviate fix location fields bug, transforming to and from geoLocation data model

kosmikdc committed:
Improves add documents and status messages for Weaviate.
disables stopword removal behavior for Weaviate

kosmikdc committed:
Weaviate query time boost fix for 5.10

kosmikdc committed:
13.2 null int field support
Weaviate ch8 improve query time boost query construction when text query is also present

kosmikdc committed:
Merge branch 'main' into weaviate_updated

kosmikdc committed:
Fixes 7.6/7.8 id inconsistency issue w/ proper id selection in cities data loader
Prevent missing product data file in chapter 8
Improve signals logging in ch6.9 (becomes minor errata)
Clean up and logs indexer.build_collection()
